internal/app/init/pkg/system/runner/containerd/containerd.go

```diff
@@ -6,147 +6,207 @@ package containerd
 
 import (
 	"context"
-	"fmt"
+	"log"
+	"path/filepath"
+	"syscall"
+	"time"
 
 	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/cio"
-	"github.com/containerd/containerd/containers"
 	"github.com/containerd/containerd/defaults"
 	"github.com/containerd/containerd/namespaces"
 	"github.com/containerd/containerd/oci"
-	"github.com/containerd/containerd/runtime/restart"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
 	"github.com/talos-systems/talos/internal/app/init/pkg/system/conditions"
 	"github.com/talos-systems/talos/internal/app/init/pkg/system/runner"
 	"github.com/talos-systems/talos/pkg/userdata"
 )
 
-// Containerd represents a service to be run in a container.
-type Containerd struct{}
+// containerdRunner is a runner.Runner that runs container in containerd
+type containerdRunner struct {
+	data *userdata.UserData
+	args *runner.Args
+	opts *runner.Options
 
-// WithMemoryLimit sets the linux resource memory limit field.
-func WithMemoryLimit(limit int64) oci.SpecOpts {
-	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
-		s.Linux.Resources.Memory = &specs.LinuxMemory{
-			Limit: &limit,
-			// DisableOOMKiller: &disable,
-		}
-		return nil
-	}
+	stop    chan struct{}
+	stopped chan struct{}
```

Comment:
It seems like it'd be worthwhile to track state changes. Should we change these chans to `time.Time` so we can save the timestamp as a new field ( `lastUpdated` for example )?

Reply:
my ultimate goal is to extract 'restart' part of the runners into a runner of its own. 

after this change both process & containerd runners have exactly same interface, and I want to extract restart into separate abstract runner which in turn accepts another runner as parameter. This way we can expose restart stats (like lastUpdated, number of restarts, current state, etc.) on that level and common to all the runners

Reply:
so this `stop` would go away, and containerd & process runners will do 'run once' semantics